File: exercises/day23-Portal-Stabilization/java/src/test/java/craft/portal/PortalStabilizerTest.java
package craft.portal;

import org.junit.jupiter.api.Test;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class PortalStabilizerTest {
    
    @Test
    void shouldProcessSingleAnomaly() {
        PortalStabilizer stabilizer = new PortalStabilizer();
        TemporalAnomaly anomaly = TemporalAnomaly.create(1000L, "alpha", 5);
        
        StabilizationReport report = stabilizer.stabilize(List.of(anomaly));
        
        assertThat(report.processedCount()).isEqualTo(1);
        assertThat(report.karmaBalances()).hasSize(1);
        assertThat(report.karmaBalances().get(0).dimension()).isEqualTo("alpha");
    }
    
    @Test
    void shouldRemoveDuplicateKarmaPatternsInSameBatch() {
        PortalStabilizer stabilizer = new PortalStabilizer();
        TemporalAnomaly anomaly1 = TemporalAnomaly.create(1000L, "alpha", 5);
        TemporalAnomaly anomaly2 = TemporalAnomaly.create(1000L, "alpha", 5); // Same karma pattern
        TemporalAnomaly anomaly3 = TemporalAnomaly.create(2000L, "beta", 3);
        
        StabilizationReport report = stabilizer.stabilize(List.of(anomaly1, anomaly2, anomaly3));
        
        assertThat(report.processedCount()).isEqualTo(2);
    }
    
    @Test
    void shouldRemoveDuplicateKarmaPatternsAcrossBatches() {
        PortalStabilizer stabilizer = new PortalStabilizer();
        TemporalAnomaly anomaly1 = TemporalAnomaly.create(1000L, "alpha", 5);
        
        stabilizer.stabilize(List.of(anomaly1));
        
        TemporalAnomaly anomaly2 = TemporalAnomaly.create(1000L, "alpha", 5); // Same karma pattern
        TemporalAnomaly anomaly3 = TemporalAnomaly.create(2000L, "beta", 3);
        
        StabilizationReport report = stabilizer.stabilize(List.of(anomaly2, anomaly3));
        
        assertThat(report.processedCount()).isEqualTo(1); // Only anomaly3 is new
        assertThat(stabilizer.getTotalProcessedCount()).isEqualTo(2); // Total: anomaly1 + anomaly3
    }
    
    @Test
    void shouldCalculateKarmaBalancePerDimension() {
        PortalStabilizer stabilizer = new PortalStabilizer();
        List<TemporalAnomaly> anomalies = List.of(
            TemporalAnomaly.create(1000L, "alpha", 10),
            TemporalAnomaly.create(2000L, "alpha", 20),
            TemporalAnomaly.create(3000L, "beta", 5)
        );
        
        StabilizationReport report = stabilizer.stabilize(anomalies);
        
        assertThat(report.karmaBalances()).hasSize(2);
        
        KarmaBalance alphaBalance = report.karmaBalances().stream()
            .filter(b -> b.dimension().equals("alpha"))
            .findFirst()
            .orElseThrow();
        
        assertThat(alphaBalance.totalKarmaDebt()).isEqualTo(30);
        assertThat(alphaBalance.anomalyCount()).isEqualTo(2);
        assertThat(alphaBalance.averageKarmaDebt()).isEqualTo(15.0);
    }
    
    @Test
    void shouldCalculateKarmaHarmony() {
        PortalStabilizer stabilizer = new PortalStabilizer();
        List<TemporalAnomaly> anomalies = List.of(
            TemporalAnomaly.create(1000L, "alpha", 2),
            TemporalAnomaly.create(2000L, "beta", 3)
        );
        
        StabilizationReport report = stabilizer.stabilize(anomalies);
        
        // Average karma debt across dimensions: (2 + 3) / 2 = 2.5
        // Karma harmony: 100 - (2.5 * 10) = 75.0
        assertThat(report.karmaHarmony()).isEqualTo(75.0);
    }
}

File: exercises/day23-Portal-Stabilization/typescript/src/tests/PortalStabilizer.test.ts
import { PortalStabilizer } from '../PortalStabilizer';
import { createTemporalAnomaly } from '../models/TemporalAnomaly';

describe('PortalStabilizer', () => {
  test('should process single anomaly', () => {
    const stabilizer = new PortalStabilizer();
    const anomaly = createTemporalAnomaly(1000, 'alpha', 5);

    const report = stabilizer.stabilize([anomaly]);

    expect(report.processedCount).toBe(1);
    expect(report.karmaBalances).toHaveLength(1);
    expect(report.karmaBalances[0].dimension).toBe('alpha');
  });

  test('should remove duplicate karma patterns in same batch', () => {
    const stabilizer = new PortalStabilizer();
    const anomaly1 = createTemporalAnomaly(1000, 'alpha', 5);
    const anomaly2 = createTemporalAnomaly(1000, 'alpha', 5); // Same karma pattern
    const anomaly3 = createTemporalAnomaly(2000, 'beta', 3);

    const report = stabilizer.stabilize([anomaly1, anomaly2, anomaly3]);

    expect(report.processedCount).toBe(2);
  });

  test('should remove duplicate karma patterns across batches', () => {
    const stabilizer = new PortalStabilizer();
    const anomaly1 = createTemporalAnomaly(1000, 'alpha', 5);

    stabilizer.stabilize([anomaly1]);

    const anomaly2 = createTemporalAnomaly(1000, 'alpha', 5); // Same karma pattern
    const anomaly3 = createTemporalAnomaly(2000, 'beta', 3);

    const report = stabilizer.stabilize([anomaly2, anomaly3]);

    expect(report.processedCount).toBe(1); // Only anomaly3 is new
    expect(stabilizer.getTotalProcessedCount()).toBe(2); // Total: anomaly1 + anomaly3
  });

  test('should calculate karma balance per dimension', () => {
    const stabilizer = new PortalStabilizer();
    const anomalies = [
      createTemporalAnomaly(1000, 'alpha', 10),
      createTemporalAnomaly(2000, 'alpha', 20),
      createTemporalAnomaly(3000, 'beta', 5),
    ];

    const report = stabilizer.stabilize(anomalies);

    expect(report.karmaBalances).toHaveLength(2);

    const alphaBalance = report.karmaBalances.find(
      (b) => b.dimension === 'alpha'
    )!;
    expect(alphaBalance.totalKarmaDebt).toBe(30);
    expect(alphaBalance.anomalyCount).toBe(2);
    expect(alphaBalance.totalKarmaDebt / alphaBalance.anomalyCount).toBe(15.0);
  });

  test('should calculate karma harmony', () => {
    const stabilizer = new PortalStabilizer();
    const anomalies = [
      createTemporalAnomaly(1000, 'alpha', 2),
      createTemporalAnomaly(2000, 'beta', 3),
    ];

    const report = stabilizer.stabilize(anomalies);

    // Average karma debt across dimensions: (2 + 3) / 2 = 2.5
    // Karma harmony: 100 - (2.5 * 10) = 75.0
    expect(report.karmaHarmony).toBe(75.0);
  });
});
